File: examples/go/clients/mcp/simple.go
package main

import (
	"context"
	"fmt"
	"os"

	x402 "github.com/coinbase/x402/go"
	evm "github.com/coinbase/x402/go/mechanisms/evm/exact/client"
	evmsigners "github.com/coinbase/x402/go/signers/evm"
	"github.com/coinbase/x402/go/mcp"
	"github.com/joho/godotenv"
	mcpsdk "github.com/modelcontextprotocol/go-sdk/mcp"
)

// MCP Client with x402 Payment Support - Simple Example
//
// This example demonstrates the RECOMMENDED way to create an MCP client
// using NewX402MCPClient with an explicitly created x402 payment client.
//
// Run with: go run . simple

// runSimple demonstrates the simple API using NewX402MCPClient with x402Client.
func runSimple() error {
	fmt.Println("\n📦 Using SIMPLE API (CreateX402MCPClient factory) with REAL MCP SDK\n")

	// Load environment variables
	if err := godotenv.Load(); err != nil {
		fmt.Println("No .env file found, using environment variables")
	}

	evmPrivateKey := os.Getenv("EVM_PRIVATE_KEY")
	if evmPrivateKey == "" {
		return fmt.Errorf("EVM_PRIVATE_KEY environment variable is required")
	}

	serverURL := os.Getenv("MCP_SERVER_URL")
	if serverURL == "" {
		serverURL = "http://localhost:4022"
	}

	fmt.Printf("🔌 Connecting to MCP server at: %s\n", serverURL)

	// Create EVM signer
	evmSigner, err := evmsigners.NewClientSignerFromPrivateKey(evmPrivateKey)
	if err != nil {
		return fmt.Errorf("failed to create EVM signer: %w", err)
	}

	fmt.Printf("💳 Using wallet: %s\n", evmSigner.Address())

	// ========================================================================
	// Connect to REAL MCP server using SSE transport
	// ========================================================================
	ctx := context.Background()

	sseClientTransport := &mcpsdk.SSEClientTransport{
		Endpoint: serverURL + "/sse",
	}

	mcpClient := mcpsdk.NewClient(&mcpsdk.Implementation{
		Name:    "x402-mcp-client",
		Version: "1.0.0",
	}, nil)

	clientSession, err := mcpClient.Connect(ctx, sseClientTransport, nil)
	if err != nil {
		return fmt.Errorf("failed to connect to MCP server: %w", err)
	}
	defer clientSession.Close()

	fmt.Println("✅ Connected to MCP server\n")

	// Create x402 payment client and wrap session
	paymentClient := x402.Newx402Client()
	paymentClient.Register("eip155:84532", evm.NewExactEvmScheme(evmSigner, nil))
	x402Mcp := mcp.NewX402MCPClient(clientSession, paymentClient, mcp.Options{
		AutoPayment: mcp.BoolPtr(true),
		OnPaymentRequested: func(context mcp.PaymentRequiredContext) (bool, error) {
			price := context.PaymentRequired.Accepts[0]
			fmt.Printf("\n💰 Payment required for tool: %s\n", context.ToolName)
			fmt.Printf("   Amount: %s (%s)\n", price.Amount, price.Asset)
			fmt.Printf("   Network: %s\n", price.Network)
			fmt.Println("   Approving payment...\n")
			return true, nil
		},
	})

	// List tools
	fmt.Println("📋 Discovering available tools...")
	toolsResult, err := clientSession.ListTools(ctx, nil)
	if err != nil {
		return fmt.Errorf("failed to list tools: %w", err)
	}

	fmt.Println("Available tools:")
	for _, tool := range toolsResult.Tools {
		fmt.Printf("   - %s: %s\n", tool.Name, tool.Description)
	}
	fmt.Println()

	// Test free tool
	fmt.Println("━" + string(make([]byte, 50)) + "━")
	fmt.Println("🆓 Test 1: Calling free tool (ping)")
	fmt.Println("━" + string(make([]byte, 50)) + "━")

	pingResult, err := x402Mcp.CallTool(ctx, "ping", map[string]interface{}{})
	if err != nil {
		return fmt.Errorf("failed to call ping: %w", err)
	}

	if len(pingResult.Content) > 0 {
		fmt.Printf("Response: %s\n", pingResult.Content[0].Text)
	}
	fmt.Printf("Payment made: %v\n\n", pingResult.PaymentMade)

	// Test paid tool
	fmt.Println("━" + string(make([]byte, 50)) + "━")
	fmt.Println("💰 Test 2: Calling paid tool (get_weather)")
	fmt.Println("━" + string(make([]byte, 50)) + "━")

	weatherResult, err := x402Mcp.CallTool(ctx, "get_weather", map[string]interface{}{
		"city": "San Francisco",
	})
	if err != nil {
		return fmt.Errorf("failed to call get_weather: %w", err)
	}

	if len(weatherResult.Content) > 0 {
		fmt.Printf("Response: %s\n", weatherResult.Content[0].Text)
	}
	fmt.Printf("Payment made: %v\n", weatherResult.PaymentMade)

	if weatherResult.PaymentResponse != nil {
		fmt.Println("\n📦 Payment Receipt:")
		fmt.Printf("   Success: %v\n", weatherResult.PaymentResponse.Success)
		if weatherResult.PaymentResponse.Transaction != "" {
			fmt.Printf("   Transaction: %s\n", weatherResult.PaymentResponse.Transaction)
		}
	}

	fmt.Println("\n✅ Demo complete!")
	return nil
}

File: examples/typescript/clients/mcp/simple.ts
/**
 * MCP Client with x402 Payment Support - Simple Example
 *
 * This example demonstrates the RECOMMENDED way to create an MCP client
 * using the high-level `createx402MCPClient` factory function.
 *
 * Run with: pnpm dev
 */

import { config } from "dotenv";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { ExactEvmScheme } from "@x402/evm/exact/client";
import { UptoEvmScheme } from "@x402/evm/upto/client";
import { createx402MCPClient } from "@x402/mcp";
import { privateKeyToAccount } from "viem/accounts";

config();

const evmPrivateKey = process.env.EVM_PRIVATE_KEY as `0x${string}`;
if (!evmPrivateKey) {
  console.error("❌ EVM_PRIVATE_KEY environment variable is required");
  process.exit(1);
}

const serverUrl = process.env.MCP_SERVER_URL || "http://localhost:4022";

/**
 * Demonstrates the simple API using createx402MCPClient factory.
 *
 * @returns Promise that resolves when demo is complete
 */
export async function main(): Promise<void> {
  console.log("\n📦 Using SIMPLE API (createx402MCPClient factory)\n");
  console.log("🔌 Connecting to MCP server at:", serverUrl);

  const evmSigner = privateKeyToAccount(evmPrivateKey);
  console.log("💳 Using wallet:", evmSigner.address);

  // ========================================================================
  // SIMPLE: One-liner setup with factory function
  // ========================================================================
  const x402Mcp = createx402MCPClient({
    name: "x402-mcp-client-demo",
    version: "1.0.0",
    schemes: [
      { network: "eip155:84532", client: new ExactEvmScheme(evmSigner) },
      { network: "eip155:84532", client: new UptoEvmScheme(evmSigner) },
    ],
    autoPayment: true,
    onPaymentRequested: async context => {
      const price = context.paymentRequired.accepts[0];
      console.log(`\n💰 Payment required for tool: ${context.toolName}`);
      console.log(`   Amount: ${price.amount} (${price.asset})`);
      console.log(`   Network: ${price.network}`);
      console.log(`   Approving payment...\n`);
      return true;
    },
  });

  // Connect using passthrough method
  const transport = new SSEClientTransport(new URL(`${serverUrl}/sse`));
  await x402Mcp.connect(transport);
  console.log("✅ Connected to MCP server\n");

  // List tools using passthrough method
  console.log("📋 Discovering available tools...");
  const tools = await x402Mcp.listTools();
  console.log("Available tools:");
  for (const tool of tools.tools) {
    console.log(`   - ${tool.name}: ${tool.description}`);
  }
  console.log();

  // Test free tool
  console.log("━".repeat(50));
  console.log("🆓 Test 1: Calling free tool (ping)");
  console.log("━".repeat(50));

  const pingResult = await x402Mcp.callTool("ping");
  console.log("Response:", pingResult.content[0]?.text);
  console.log("Payment made:", pingResult.paymentMade);
  console.log();

  // Test paid tool
  console.log("━".repeat(50));
  console.log("💰 Test 2: Calling paid tool (get_weather)");
  console.log("━".repeat(50));

  const weatherResult = await x402Mcp.callTool("get_weather", { city: "San Francisco" });
  console.log("Response:", weatherResult.content[0]?.text);
  console.log("Payment made:", weatherResult.paymentMade);

  if (weatherResult.paymentResponse) {
    console.log("\n📦 Payment Receipt:");
    console.log("   Success:", weatherResult.paymentResponse.success);
    if (weatherResult.paymentResponse.transaction) {
      console.log("   Transaction:", weatherResult.paymentResponse.transaction);
    }
  }

  console.log("\n✅ Demo complete!");
  await x402Mcp.close();
  process.exit(0);
}
